### PR TITLE
feat: add matrix inversion

### PR DESCRIPTION
**Pull request by polvalente**
Adds matrix inversion

Indirectly related to #157 and unblocks some of the grads in #161

### CODE REVIEW

**Comment by seanmor5** [comment]
Hi, @polvalente! This looks good, I am only hesitant to add this into the API because it looks pretty easily implementable as a routine in defn, but it is probably useful if it is going to be a common method. I don't work much with matrix inversion or any of the other linalg operators for that matter that often, so I can't speak much on how often it will be used. Are there a lot of use cases?

It might also be time for a discussion on how we are going to split out other linear algebra operators, random operators, etc. that are built on top of our existing primitives.

**Comment by polvalente** [comment]
> Hi, @polvalente! This looks good, I am only hesitant to add this into the API because it looks pretty easily implementable as a routine in defn, but it is probably useful if it is going to be a common method. I don't work much with matrix inversion or any of the other linalg operators for that matter that often, so I can't speak much on how often it will be used. Are there a lot of use cases?
> 
> It might also be time for a discussion on how we are going to split out other linear algebra operators, random operators, etc. that are built on top of our existing primitives.

My main motivation for this is because matrix inversion is needed for some of the grads I explored throughout this week (namely cholesky, QR and LU). Since this was basically fully implemented using the existing API, I'm ok with having it as a private function in the Nx.Grad module.

However, other people might find this useful. For instance, we can use inversion for reversing computer graphics operations (i.e. combined rotation and transalation matrices). 

As you said, maybe we could have this as a part of a documented linalg tutorial which could be implemented by users as needed. The comments included in this also show how to use QR together with triangular_solve for solving any (sovable) linear system.

Another approach we could take with inversion specifically is to turn this into a binary backend implementation so we can delegate this to the proper implementation in other backends.

File: nx/lib/nx.ex [hunk 1]
```diff
@@ -7919,6 +7921,101 @@ defmodule Nx do
...
+  @doc type: :linalg
+  def invert(tensor) do
+    %T{shape: {m, n}} = tensor = tensor!(tensor)
+
+    if m != n,
+      do:
+        raise(
+          ArgumentError,
+          "can only invert square matrices, got: {#{m}, #{n}}"
+        )
```
Comment (josevalim) [inline nx/lib/nx.ex:7981]: This is unnecessarily verbose in this format. Maybe it should be a regular do block?

```suggestion
    if m != n do
      raise ArgumentError,
              "can only invert square matrices, got: {#{m}, #{n}}"
    end
```

File: nx/lib/nx.ex [hunk 2]
```diff
@@ -7919,6 +7921,101 @@ defmodule Nx do
...
+    |> Enum.map(fn i ->
+      b = identity |> slice([0, i], [n, 1]) |> Nx.reshape({n})
+
+      y =
+        try do
+          triangular_solve(l, b)
+        rescue
+          ArgumentError ->
+            raise ArgumentError, "can't invert a singular matrix"
+        end
```
Comment (josevalim) [inline nx/lib/nx.ex:8012]: This try/catch won't work when invoking inverse inside a `defn` so perhaps we should remove it? :) What is its purpose?

Comment (polvalente) [reply nx/lib/nx.ex:8012]: Better error messages. Triangular solve will return something like "can't solve singular system", which isn't the most semantic message, but this isn't necessary for the algorithm itself

**Comment by josevalim** [comment]
Hi @polvalente, in JAX they implement inverse on top of triangular_solve. This is because both numpy/jax allow triangular_solve for a and b having the same rank. Here is the code: https://github.com/google/jax/blob/f8c36d9c0a0920c5d68f578cbf3e05d3a5aaa182/jax/_src/numpy/linalg.py#L331-L336

My proposal to add inverse therefore is to solve this using `triangular_solve`. `triangular_solve` is most likely optimized by almost all backends which will make sure this is efficient.

WDYT?

**Comment by polvalente** [comment]
> Hi @polvalente, in JAX they implement inverse on top of triangular_solve. This is because both numpy/jax allow triangular_solve for a and b having the same rank. Here is the code: https://github.com/google/jax/blob/f8c36d9c0a0920c5d68f578cbf3e05d3a5aaa182/jax/_src/numpy/linalg.py#L331-L336
> 
> My proposal to add inverse therefore is to solve this using `triangular_solve`. `triangular_solve` is most likely optimized by almost all backends which will make sure this is efficient.
> 
> WDYT?

Sounds better! It seems they use `solve` but we can use all these tricks I've used in inverting for solving a general `A . B = C` system in binary backend (which in turn would use `triangular_solve`), delegating `solve` to the proper implementations in other backends.

IMO we should also use this opportunity to expose `solve` as part of the API, since it's a more general case then `triangular_solve`. 

Other than that, this also improves the AST in `Expr`, because using this implementation the code actually shows the inner workings of `invert` instead of showing `invert a` as a step.

How does that sound?

File: nx/lib/nx.ex (hunk 2 above)
Comment (josevalim) [reply nx/lib/nx.ex:8010]: Ok. The issue is that this error message won't work inside `defn`, so perhaps we should improve the error message in the BinaryBackend if possible.

**Comment by josevalim** [comment]
Ah, you are correct, `inverse` is using solve and not triangular_solve. I mixed them up. In this case, I would not expose `solve` for now and go ahead with `inverse`. Also note their `solve` stops computing the gradient [for the input here](https://github.com/google/jax/blob/f8c36d9c0a0920c5d68f578cbf3e05d3a5aaa182/jax/_src/lax/linalg.py#L264). We may need to do the same to compute the gradient of inverse but I am not sure. Perhaps we need some tests on the gradient of inverse. :)

File: nx/lib/nx.ex (hunk 2 above)
Comment (AlexJuca) [reply nx/lib/nx.ex:8010]: @josevalim Just for curiosity, what are current limitations of `defn` and what do we expect to have solved in terms of limitations by the time the API becomes stable?

File: nx/lib/nx.ex [hunk 3]
```diff
@@ -7903,22 +7919,90 @@ defmodule Nx do
...
-      raise ArgumentError, "incompatible dimensions for a and b on tringular solve"
+    case b_shape do
+      {^m, ^m} ->
+        nil
+
+      {^m} ->
+        nil
+
+      _ ->
+        raise ArgumentError, "incompatible dimensions for a and b on tringular solve"
```
Comment (AlexJuca) [inline nx/lib/nx.ex:7937]: I believe this should be "triangular" and not "tringular"

Comment (AlexJuca) [reply nx/lib/nx.ex:7937]:
```suggestion
        raise ArgumentError, "incompatible dimensions for a and b on triangular solve"
```

**Review by AlexJuca — CHANGES_REQUESTED** [review]

File: nx/lib/nx.ex (hunk 3 above)
Comment (polvalente) [reply nx/lib/nx.ex:7937]: yeap

File: nx/lib/nx/defn/expr.ex [hunk 4]
```diff
@@ -643,6 +643,13 @@ defmodule Nx.Defn.Expr do
     composite({u, s, vt}, context, &expr(&1, context, :svd, [{u, s, vt}, tensor, opts]))
   end
 
+  @impl true
+  def invert(output, input) do
+    input = to_expr(input)
+    context = input.data.context
+    composite(output, context, &expr(&1, context, :invert, [output, input]))
```
Comment (josevalim) [inline nx/lib/nx/defn/expr.ex:650]: No need to call composite as it never returns a tuple:

```suggestion
    expr(output, context, :invert, [output, input])
```

File: nx/lib/nx/binary_backend.ex [hunk 5]
```diff
@@ -1115,15 +1117,52 @@ defmodule Nx.BinaryBackend do
...
+    y = triangular_solve(l, l, b, [])
+
+    dot(output_holder, qt, [1], y, [0])
+  end
+
+  @impl true
+  def invert(output, %T{shape: {n, n}} = tensor) do
+    identity = eye(tensor, [])
+    solve(output, tensor, identity)
+  end
```
Comment (josevalim) [inline nx/lib/nx/binary_backend.ex:1164]: Given both solve and invert work exclusively on existing operations, I think we can move this implementations to Nx. WDYT? This way we don't add any new function to the backend. :)

File: nx/lib/nx/binary_backend.ex [hunk 6]
```diff
@@ -1115,15 +1117,52 @@ defmodule Nx.BinaryBackend do
...
+      when b_shape == {rows, rows} or b_shape == {rows} do
     a_data = to_binary(a)
     b_data = to_binary(b)
-    out_bin = B.Decomposition.ts(a_data, a_type, b_data, b_type, rows, output_type, opts)
+    out_bin = B.Decomposition.ts(a_data, a_type, b_data, b_type, b_shape, output_type, opts)
     from_binary(out, out_bin)
   end
 
+  @doc false
+  def solve(%T{type: output_type} = output_holder, %T{shape: {n, n}} = a, %T{shape: b_shape} = b)
```
Comment (josevalim) [inline nx/lib/nx/binary_backend.ex:1131]: Let's also expose this on `Nx` as `solve`. :) And given it works on existing tensors, there is no need for a backend function as well!

**Review by josevalim — COMMENTED** [review]
This looks great! I have dropped some minor comments about code organization and we can ship it.

**Comment by josevalim** [comment]
I tried to resist but we should probably have our own `Nx.LinAlg` namespace. :) We can focus on that after this PR is merged.

**Comment by AlexJuca** [comment]
@polvalente @josevalim I would love to handle the move to the `Nx.LinAlg` namespace, after these changes are merged.

File: nx/lib/nx/binary_backend.ex (hunk 5 above)
Comment (polvalente) [reply nx/lib/nx/binary_backend.ex:1164]: For this function specifically I think we can benefit of existing implementations on XLA and other backends. Matrix inversion has lots of optimized versions (with divide and conquer algorithms and such), so I'd rather delegate it to the low-level backends whenever possible

File: nx/lib/nx/binary_backend.ex (hunk 6 above)
Comment (polvalente) [reply nx/lib/nx/binary_backend.ex:1131]: This has the same issue as invert, albeit in a milder way. There are some algorithms which can be more efficient for solving non-triangularized matrices.

If we're exposing `solve`, I'd rather expose it delegating the implementation to each backend. Are there any major downsides in doing this?

**Comment by josevalim** [comment]
@polvalente i know backends can optimize those implementations but we don’t quite support it yet. So let’s move them to Nx and we will add optional callbacks to the future. The biggest upside right now of implementing them on Nx is that we get grad for free.

**Comment by polvalente** [comment]
> @polvalente i know backends can optimize those implementations but we don’t quite support it yet. So let’s move them to Nx and we will add optional callbacks in the future. The biggest upside right now of implementing them on Nx is that we get grad for free.

I see! On it, then :)

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**